mce_listen.8
.TH mce_listen ""
.\" Copyright (c) 2007 Tim Hockin (thockin@google.com)
.SH NAME
mce_listen \- MCE event listener
.SH SYNOPSIS
\fBmce_listen\fP [\fIoptions\fP]

.SH DESCRIPTION
\fBmced\fP is the sysem-wide MCE event catcher.  \fBmce_listen\fP is a
simple shell-friendly tool which connects to \fBmced\fP and listens for
events.  When an event occurs, \fBmce_listen\fP will print it on stdout.

.SH OPTIONS
.TP
.BI \-c "\fR, \fP" \--count " events"
Receive up to a maximum number of MCE events, then exit.  Specifying 0
events here will cause \fBmce_listen\fP to open the connection and
immediately exit.  This is useful as a 'ping' of sorts.
.TP
.BI \-s "\fR, \fP" \--socketfile " filename"
This option changes the name of the UNIX domain socket which
\fBmce_listen\fP opens.  Default is \fI/var/run/mced2.socket\fP.
.TP
.BI \-O "\fR, \fP" \--oldsocket
This option forces \fBmce_listen\fP to use the legacy socket path
(\fI/var/run/mced.socket\fP).  The \-s (\--socketfile) flag always
takes precedence over this flag.
.TP
.BI \-t "\fR, \fP" \--time " seconds"
Listen for the specified time in seconds, before exiting.  Setting this to
0 or less will cause \fBmce_listen\fP to listen with no timeout.  Default
is no timeout.
.TP
.BI \--dbus
Listen for events on D-Bus rather than the UNIX socket.  Default is to use
the UNIX socket.
.TP
.BI \--dbus-session-bus
Listen for D-Bus signals on the session bus, rather than the system bus.
Default is the system bus.
.TP
.BI \-v "\fR, \fP" \--version
Print version information and exit.
.TP
.BI \-h "\fR, \fP" \--help
Show help and exit.

.SH FILES
.PD 0
.B /var/run/mced2.socket
.br
.B /var/run/mced.socket
.PD

.SH BUGS
There are no known bugs.  To file bug reports, see \fBAUTHORS\fP below.
.SH SEE ALSO
regcomp(3), sh(1), socket(2), connect(2)
.SH AUTHORS
Tim Hockin <thockin@hockin.org>
.br



Changelog
%changelog
* Sat Feb 27 2010  Tim Hockin <thockin@google.com>
  - Rename examples to be valid config files.

* Fri Feb 26 2010  Tim Hockin <thockin@google.com>
  - Make config file format more flexible wrt whitespace.

* Thu Feb 25 2010  Tim Hockin <thockin@google.com>
  - Rename some 'struct mce' fields to accommodate new fields better.
  - Added CPU socket as %S.
  - Added CPU initial APIC ID as %p.
  - Added MCG_CAP as %G.
  - Send %T (tsc), %C (cs), and %I (ip) on the socket and DBUS, for
    completeness.
  - Fix bug in client rules that caused an extra newline to be sent.
  - Update mce_decode with new struct field names.
  - Use mced2.socket instead of mced.socket by default.

* Sun Feb 21 2010  Tim Hockin <thockin@hockin.org>
  - Add support for NULL short or long options. [cmdline.c]
  - Add D-Bus support.  [mced.c, mced.8, mce_listen.c, mce_listen.8,
    dbus.c, dbus.h, dbus_asv.h, dbus_asv.c, dbus_interface.xml ]
  - Add CPU vendor as %v. [mced.c, mced.h, mced.8, rules.c, dbus.c]
  - Add CPUID 1 EAX as %A. [mced.c, mced.h, mced.8, rules.c, dbus.c]

* Wed Feb 17 2010  Tim Hockin <thockin@google.com>
  - Support linux-2.6.31's new larger struct mce. [mced.c, mced.h, mced.8]

* Mon Feb 15 2010  Tim Hockin <thockin@google.com>
  - Fix numerous bugs and typos in man page. [mced.8]
  - Change socket output to a new "%x=value" format.  It's more extensible.
    Document it better and deprecate the old style. [mced.8, rules.c]
  - Add a -O (--oldsocket) flag to retain the old socket data format for
    compatibility. [mced.c, mced.h, rules.c]
  - Prep to make this be version 2.0.0. [Makefile]
  - Convert mce_list to use cmdline.h. [mce_listen.c, mce_listen.8, Makefile]

* Sun Feb 14 2010  Tim Hockin <thockin@google.com>
  - Use my own cmdline library, rather than getopt.  It's less fragile and
    less likely to get screwed up. [mced.c, mced.8, cmdline.c, cmdline.h,
    Makefile]
  - Change order of commandline flags in usage output. [mced.c]

* Fri Feb 12 2010  Tim Hockin <thockin@google.com>
  - Add -o (--oflowsuppress) flag to manage overflow log message
    suppression.  [mced.h, mced.c, mced.8]

* Wed Oct 14 2009  Tim Hockin <thockin@google.com>
  - Factor out kernel MCE record length in prep for 2.6.31 compatibility
    updates.  [mced.h, mced.c]

* Tue Oct 13 2009  Tim Hockin <thockin@google.com>
  - Fix 64-bit print warnings. [rules.c, mce_decode.c]

* Wed Sep 02 2009  Tim Hockin <thockin@google.com>
  - Fix small regression in logging behavior wrt client connections. [rules.c]

* Tue Aug 11 2009  Tim Hockin <thockin@google.com>
  - Clean up debug logging to not emit messages unless -d is explicitly
    specified.  [mced.c, mced.h, rules.c]

* Fri May 15 2009  Tim Hockin <thockin@hockin.org>
  - Make client connect/disconnect log messages be debug-only. [rules.c]

* Thu May 07 2009  Tim Hockin <thockin@hockin.org>
  - Convert all fprintf() calls to mced_log(). [mced.c]
  - Add -R (--retrydev) option to make mced retry /dev/mcelog if it fails
    to open. [mced.c, mced.8]

* Mon May 04 2009  Tim Hockin <thockin@hockin.org>
  - Make mce_listen accept '-c 0' as a ping. [mce_listen.c mce_listen.8]
  - Log a message when we take a clean_exit signal. [mced.c]

* Sat May 02 2009  Tim Hockin <thockin@hockin.org>
  - Bump to version 1.0.0 for release. [Makefile]
  - Bump to version 1.0.1 after release. [Makefile]
  - Fix libdb linking erroneously when not needed. [Makefile]

* Mon Mar 07 2009  Tim Hockin <thockin@hockin.org>
  - Change default socket mode to 0600. [mced.c]
  - Add a -C (--clientmax) command line flag to set max number of non-root
    socket connections. [mced.c rules.c mced.h]
  - Set the maximum number of socket clients to 256 by default. [mced.h]
  - Close clients that have disconnected. [mced.c rules.c]
  - Give up and exit() if 5 accept() calls fail in a row. [mced.c]

* Mon Feb 09 2009  Tim Hockin <thockin@hockin.org?
  - Open /dev/null O_RDWR, rather than O_RDONLY. [mced.c]

* Thu Jan 08 2009  Tim Hockin <thockin@hockin.org>
  - Fix help string - -d does not imply -l. [mced.c]

* Tue Dec 23 2008  Tim Hockin <thockin@hockin.org>
  - Add generic rate limiting support, used for some log messages. [mced.c]
  - Use MCi_STATUS.OVER to report overflows. Rate limit this to once per
    10 seconds. [mced.c, mced.h]
  - Check the flags from mcelog, and report overflows. [mced.c, mced.h]
  - Fix a small potential leak in the error path of open_log. [mced.c]
  - Change the socketpath to be in /var/run, rather than a specific subdir
    thereof. [mced.c]
  - Add log messages if the daemon fails to start up properly. [mced.c]

* Mon Dec 22 2008  Tim Hockin <thockin@hockin.org>
  - Fix some wording in the man page wrt rate limiting. [mced.8]
  - Make the default bootnum be -1, rather than 0. [mced.c, mced.h, mced.8]
  - Add TSC, CS, and IP fields to struct mce from kenrel_mce. [mced.c,
    mced.h, mced.8, rules.c]
  - Fix a leak closedir() in case of malloc() failure. [rules.c]

* Sun Dec 21 2008  Tim Hockin <thockin@hockin.org>
  - Do not imply -l with -d. [mced.c, rules.c]
  - Exit if a fake mcelog device closes. [mced.c]
  - Change numeric commandline args to be long, not int. [mced.c]
  - Add rate limiting via -r (--ratelimit) for MCEs. [mced.c, mced.8]
  - Add mcelog_faker for testing. [mcelog_Faker.c]

* Sun Dec 14 2008  Tim Hockin <thockin@hockin.org>
  - Fix Changelog ordering. [Changelog]
  - Change -i (--interval) to be -x (--max_interval) and -n (--mininterval)
    options. [mced.c, mced.h, mced.8]
  - Add support for using a FIFO as /dev/mcelog for testing.  Compiled off
    by default. This require O_NONBLOCK on the open(). [mced.c, Makefile]
  - Make mced_vlog() static. [mced.h, mced.c]
  - Move the GCC printf attribute to th header. [mced.c, mced.h]
  - Tidy up debug prints around the interval adjustment logic. [mced.c]
  - Detect if poll() timed out, and use that in the interval adjustment
    logic. [mced.c]

mce_listen.c
/*
 *  mce_listen.c - client for mced's UNIX socket
 *
 *  Based on code from acpid.
 *  Copyright (c) 2007 Tim Hockin (thockin@hockin.org)
 *  Copyright (c) 2007 Google, Inc. (thockin@google.com)
 *  Portions Copyright (c) 2004 Tim Hockin (thockin@hockin.org)
 *  Portions Copyright (c) 2001 Sun Microsystems
 *  Portions Copyright (c) 2000 Andrew Henroid
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <signal.h>
#include <unistd.h>
#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <string.h>
#include <errno.h>
#include <ctype.h>
#include <time.h>
#include <sys/poll.h>
#include <grp.h>
#include <signal.h>

#include "mced.h"
#include "util.h"
#include "cmdline.h"
#include "ud_socket.h"

#if ENABLE_DBUS
#include <glib.h>
#include "dbus.h"
#include "dbus_asv.h"
#include "auto.dbus_client.h"
#endif

static cmdline_string socketfile = NULL;
static cmdline_int max_events = -1;
static cmdline_int time_limit = -1;
static cmdline_bool use_v1_socket = 0;

#if ENABLE_DBUS
static cmdline_bool use_dbus = 0;
static cmdline_bool use_session_dbus = 0;
#endif

static void do_help(const struct cmdline_opt *, ...);
static void do_version(const struct cmdline_opt *, ...);
static struct cmdline_opt cmdline_opts[] = {
	{
		"c", "count",
		CMDLINE_OPT_INT, &max_events,
		"<num>", "Set the maximum number of events"
	},
	{
		"s", "socketfile",
		CMDLINE_OPT_STRING, &socketfile,
		"<file>", "Use the specified socket file"
	},
	{
		"O", "oldsocket",
		CMDLINE_OPT_BOOL, &use_v1_socket,
		"", "Make socket behavior compatible with mced v1.x"
	},
	{
		"t", "time",
		CMDLINE_OPT_INT, &time_limit,
		"<secs>", "Listen for the specified time (in seconds)"
	},
	#if ENABLE_DBUS
	{
		NULL, "dbus",
		CMDLINE_OPT_BOOL, &use_dbus,
		"", "Use D-Bus to listen for events"
	},
	{
		NULL, "dbus-session-bus",
		CMDLINE_OPT_BOOL, &use_session_dbus,
		"", "Use D-Bus session bus, instead of system bus"
	},
	#endif
	{
		"v", "version",
		CMDLINE_OPT_CALLBACK, do_version,
		"", "Print version information and exit"
	},
	{
		"h", "help",
		CMDLINE_OPT_CALLBACK, do_help,
		"", "Print this help message and exit"
	},
	CMDLINE_OPT_END_OF_LIST
};

static void
usage(FILE *out)
{
	const char *help_str;
	fprintf(out, "Usage: %s [OPTIONS]\n", cmdline_progname);
	fprintf(out, "\n");
	while ((help_str = cmdline_help(cmdline_opts))) {
		fprintf(out, "  %s\n", help_str);
	}
	fprintf(out, "\n");
}

/*
 * Parse command line arguments
 */
static int
handle_cmdline(int *argc, const char ***argv)
{
	/* Parse the command line. */
	if (cmdline_parse(argc, argv, cmdline_opts) != 0) {
		usage(stderr);
		exit(EXIT_FAILURE);
	}
	if (*argc != 1) {
		fprintf(stderr,
		        "Unknown command line argument: '%s'\n\n", (*argv)[1]);
		usage(stderr);
		exit(EXIT_FAILURE);
	}

	/*
	 * Post-process command line flags.
	 */
	if (max_events < 0) {
		max_events = -1;
	}
	if (time_limit > 0) {
		alarm(time_limit);
	}
	if (socketfile == NULL) {
		if (use_v1_socket) {
			socketfile = MCED_SOCKETFILE_V1;
		} else {
			socketfile = MCED_SOCKETFILE_V2;
		}
	}

	return 0;
}

static void
do_help(const struct cmdline_opt *opt __attribute__((unused)), ...)
{
	usage(stdout);
	exit(EXIT_SUCCESS);
}

static void
do_version(const struct cmdline_opt *opt __attribute__((unused)), ...)
{
	printf(PACKAGE "-" PRJ_VERSION "\n");
	exit(EXIT_SUCCESS);
}

static int
socket_client(void)
{
	int sock_fd;
	int ret;

	/* open the socket */
	sock_fd = ud_connect(socketfile);
	if (sock_fd < 0) {
		fprintf(stderr, "%s: can't open socket %s: %s\n",
			cmdline_progname, socketfile, strerror(errno));
		exit(EXIT_FAILURE);
	}
	fcntl(sock_fd, F_SETFD, FD_CLOEXEC);

	if (max_events == 0) {
		exit(EXIT_SUCCESS);
	}

	/* set stdout to be line buffered */
	setvbuf(stdout, NULL, _IOLBF, 0);

	/* main loop */
	ret = 0;
	while (1) {
		char *event;

		/* read and handle an event */
		event = read_line(sock_fd);
		if (event) {
			fprintf(stdout, "%s\n", event);
		} else if (errno == EPIPE) {
			fprintf(stderr, "connection closed\n");
			break;
		} else {
			static int nerrs;
			if (++nerrs >= MCED_MAX_ERRS) {
				fprintf(stderr, "too many errors - aborting\n");
				ret = 1;
				break;
			}
		}

		if (max_events > 0 && --max_events == 0) {
			break;
		}
	}

	return ret;
}

#if ENABLE_DBUS
/* handler for MCE events */
static void
dbus_mce_handler(DBusGProxy *proxy __attribute__((unused)),
                 const dbus_asv *asv,
                 void *data __attribute__((unused)))
{
	const char *key;
	GValue *value;
	int i = 0;

	dbus_asv_iterator it;
	dbus_asv_iter_init(&it, asv);
	while (dbus_asv_iter_next(&it, &key, &value)) {
		if (i++ != 0) {
			printf(" ");
		}
		gboolean valid;
		if (G_VALUE_HOLDS_INT(value)) {
			int32_t val = dbus_asv_get_int32(asv, key, &valid);
			if (valid) {
				printf("%s=%ld", key, (long)val);
			}
		} else if (G_VALUE_HOLDS_UINT(value)) {
			uint32_t val = dbus_asv_get_uint32(asv, key, &valid);
			if (valid) {
				printf("%s=0x%08lx", key, (unsigned long)val);
			}
		} else if (G_VALUE_HOLDS_INT64(value)) {
			int64_t val = dbus_asv_get_int64(asv, key, &valid);
			if (valid) {
				printf("%s=%lld", key, (long long)val);
			}
		} else if (G_VALUE_HOLDS_UINT64(value)) {
			uint64_t val = dbus_asv_get_uint64(asv, key, &valid);
			if (valid) {
				printf("%s=0x%016llx", key,
				       (unsigned long long)val);
			}
		}
	}
	printf("\n");

	if (max_events > 0 && --max_events == 0) {
		exit(EXIT_SUCCESS);
	}
}

/* Run a dbus mainloop */
static int
dbus_client(void)
{
	int which_bus;
	DBusGConnection *bus;
	DBusGProxy *remote_value;
	GMainLoop *main_loop;
	GError *error = NULL;

	g_type_init();

	main_loop = g_main_loop_new(NULL, FALSE);
	if (main_loop == NULL) {
		fprintf(stderr, "can't create dbus main loop");
		return -1;
	}

	if (use_session_dbus) {
		which_bus = DBUS_BUS_SESSION;
	} else {
		which_bus = DBUS_BUS_SYSTEM;
	}
	bus = dbus_g_bus_get(which_bus, &error);
	if (error != NULL) {
		fprintf(stderr, "can't connect to dbus");
		return -1;
	}

	remote_value = dbus_g_proxy_new_for_name(bus,
	                                         DBUS_SERVICE_NAME,
	                                         DBUS_SERVICE_OBJECT_PATH,
	                                         DBUS_SERVICE_INTERFACE);
	if (remote_value == NULL) {
		fprintf(stderr, "can't create the dbus proxy");
		return -1;
	}

	dbus_g_proxy_add_signal(remote_value, DBUS_SIGNAL_NAME_MCE,
	                        dbus_asv_gtype(), G_TYPE_INVALID);

	dbus_g_proxy_connect_signal(remote_value, DBUS_SIGNAL_NAME_MCE,
	                            G_CALLBACK(dbus_mce_handler), NULL, NULL);

	if (max_events == 0) {
		return EXIT_SUCCESS;
	}

	g_main_loop_run(main_loop);
	return -1;
}
#endif  /* ENABLE_DBUS */

static void
time_expired(int signum __attribute__((unused)))
{
	exit(EXIT_SUCCESS);
}

int
main(int argc, const char *argv[])
{
	/* handle an alarm */
	signal(SIGALRM, time_expired);

	/* handle the commandline  */
	handle_cmdline(&argc, &argv);

	/* listen for events */
	#if ENABLE_DBUS
	if (use_dbus) {
		return dbus_client();
	}
	#endif
	return socket_client();
}
